app/models/concerns/course/material/folder/ordering_concern.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+module Course::Material::Folder::OrderingConcern
+  extend ActiveSupport::Concern
+
+  # Sorts all folders in a collection in topological order.
+  #
+  # By convention, each folder is represented by an array. The first element is the folder itself,
+  # the second is the children of the array.
+  class FolderSort
+    include Enumerable
+    delegate :each, to: :@sorted
+    delegate :length, to: :@sorted
+    delegate :flatten, to: :@sorted
+    alias_method :size, :length
+
+    # Constructor.
+    #
+    # @param [Array<Course::Material::Folder>] folders The folders to sort.
+    def initialize(folders)
+      @folders = folders
+      @sorted = sort(nil)
+    end
+
+    # Retrieves the last folder topologically -- the last folder at every branch.
+    #
+    # @return [Course::Material::Folder] The last folder topologically.
+    # @return [nil] When there are no folders.
+    def last
+      current_thread = @sorted.last
```

Comment:
Shouldn't be called thread right? `current_folder` might be better in this context.
Also, in the comment description, "last folder topologically" means the root, or the leaf? Current phrasing leaves it ambiguous.